***
# **Customer Purchase Behavior Analysis and Reporting**
***

## **Data Extraction**

In [140]:
# Importing sqlalchemy
import sqlalchemy

In [141]:
# Installation of pymysql
!pip install pymysql

In [142]:
import pymysql

In [143]:
import pandas as pd
import numpy as np
from numpy import random

In [144]:
# Creating connection between SQL and Python
engine = sqlalchemy.create_engine('mysql+pymysql://root:2001@localhost:3306/construct_week_project')

In [145]:
#Loading customers, products, purchase tables from SQLDB
customers = pd.read_sql_table("customers",engine)
products = pd.read_sql_table("products",engine)
purchase = pd.read_sql_table("purchase",engine)
reviews = pd.read_sql_table("reviews",engine)

## **Sentiment Analysis**

In [147]:
random.seed(25)
Sentiment = random.choice(['Positive', 'Negative', 'Neutral'], size=(1000,1))
reviews = reviews.assign(Sentiment=Sentiment)
reviews

ReviewID  CustomerID  ProductID  \
0           1         882         18   
1           2         702         12   
2           3         386         18   
3           4         916         19   
4           5         200         12   
..        ...         ...        ...   
995       996         794          3   
996       997         194         15   
997       998         255         18   
998       999         696         15   
999      1000         412          1   

                                            ReviewText ReviewDate Sentiment  
0                          Just return may onto black. 2024-05-12  Positive  
1         Represent scientist none finish again truth. 2024-03-07   Neutral  
2                             Parent behind cover top. 2023-11-15   Neutral  
3    Thousand similar would still to common rather ... 2024-05-27  Negative  
4         Himself water maintain less west age notice. 2023-09-10   Neutral  
..                                                 ...        ...       ...  
995                    Worker travel agency seek dark. 2024-04-08  Positive  
996             Way management can Democrat book base. 2023-09-06   Neutral  
997                      Support know along firm hour. 2023-09-30  Positive  
998                      Defense hold exactly kitchen. 2024-03-02  Positive  
999                               Big manager man may. 2023-11-23  Positive  

[1000 rows x 6 columns]

## **Data Analysis**

#### *Statistics*

In [150]:
# Calculating total purchase quantity
Total_Purchase_Quantity = sum(purchase["PurchaseQuantity"])
Total_Purchase_Quantity

3053

In [151]:
# Calculating total purchase quantity
Total_Purchase_Quantity = sum(purchase["PurchaseQuantity"])
Total_Purchase_Quantity

3053

In [152]:
# Calculating total revenue
Total_Revenue = sum(purchase["PurchasePrice"])
Total_Revenue

489274.72

In [153]:
# Calculating average purchase value
Average_Purchase_Value = purchase["PurchasePrice"].mean()
Average_Purchase_Value

489.27471999999995

#### *Identify top customers and their purchasing behavior*

In [155]:
merge1 = pd.merge(customers, purchase, on = "CustomerID")
merge2 = pd.merge(merge1, products, on = "ProductID")
top_customers = merge2.groupby(["CustomerID", "CustomerName","PurchaseDate",
                                "ProductName", "PurchaseQuantity","ProductCategory"]).sum()["PurchasePrice"].sort_values(ascending=False)
top_customers.head(10)

CustomerID  CustomerName      PurchaseDate  ProductName     PurchaseQuantity  ProductCategory
778         Jillian Baker     2024-03-21    Vacuum Cleaner  2                 Home Appliances    999.98
268         Patricia Taylor   2023-09-26    Smartphone      5                 Electronics        999.44
282         Mikayla Rios      2023-10-04    Camera          5                 Electronics        998.17
74          Matthew Espinoza  2023-07-24    Television      3                 Electronics        997.45
504         Jamie Montoya     2023-12-21    Headphones      4                 Electronics        996.01
872         Charles Clark     2024-05-03    Smartphone      3                 Electronics        995.97
460         Antonio Hughes    2023-12-05    Vacuum Cleaner  4                 Home Appliances    994.58
927         Carlos Smith      2024-05-24    Headphones      5                 Electronics        993.55
816         Steven Rivera     2024-04-08    Refrigerator    2             

#### *Analyze purchase trends over time (monthly, quarterly, yearly)*

In [157]:
purchase['Month'] = purchase["PurchaseDate"].dt.month
purchase['Year'] = purchase["PurchaseDate"].dt.year
purchase['Quarter'] = purchase["PurchaseDate"].dt.quarter
purchase

TransactionID  CustomerID  ProductID PurchaseDate  PurchaseQuantity  \
0                1         726          1   2024-03-01                 5   
1                2         620          2   2024-01-26                 4   
2                3         904         11   2024-05-13                 3   
3                4         249          3   2023-09-21                 5   
4                5         617         12   2024-01-25                 4   
..             ...         ...        ...          ...               ...   
995            996         552         11   2024-01-03                 2   
996            997         385         10   2023-11-11                 5   
997            998         364         15   2023-11-04                 5   
998            999         366          5   2023-11-05                 4   
999           1000         925          4   2024-05-22                 5   

     PurchasePrice  Month  Year  Quarter  
0           689.99      3  2024        1  
1            79.27      1  2024        1  
2           666.75      5  2024        2  
3           316.19      9  2023        3  
4           700.24      1  2024        1  
..             ...    ...   ...      ...  
995         796.67      1  2024        1  
996         271.88     11  2023        4  
997         594.69     11  2023        4  
998         140.89     11  2023        4  
999         861.00      5  2024        2  

[1000 rows x 9 columns]

##### 1. Monthly

In [159]:
avg_monthly_sales = purchase.groupby("Month")["PurchasePrice"].mean()
avg_monthly_sales

Month
1     449.228247
2     486.985529
3     542.647246
4     517.661212
5     509.523452
6     471.103134
7     515.663333
8     434.980000
9     502.847312
10    498.298312
11    419.360947
12    535.158969
Name: PurchasePrice, dtype: float64

##### 2. Quarterly

In [161]:
avg_quarterly_sales = purchase.groupby("Quarter")["PurchasePrice"].mean()
avg_quarterly_sales

Quarter
1    487.695498
2    500.136037
3    487.509316
4    483.712565
Name: PurchasePrice, dtype: float64

##### 3. Yearly

In [163]:
avg_yearly_sales = purchase.groupby("Year")["PurchasePrice"].mean()
avg_yearly_sales

Year
2023    483.141934
2024    496.561597
Name: PurchasePrice, dtype: float64

#### *Identify the top-performing product categories*

In [165]:
merge3 = pd.merge(purchase, products, on = "ProductID")
Category_Sales = merge3.groupby("ProductCategory")["PurchasePrice"].sum().sort_values(ascending=False)
Category_Sales

ProductCategory
Electronics        248194.56
Home Appliances    241080.16
Name: PurchasePrice, dtype: float64

#### *Identify the top-performing products*

In [167]:
Product_Sales = merge3.groupby("ProductName")["PurchasePrice"].sum().sort_values(ascending=False)
Product_Sales.head(10)

ProductName
Microwave          30970.75
Smartwatch         29407.68
Toaster            29116.03
Heater             28828.67
Headphones         27748.78
Refrigerator       27730.95
Printer            27360.29
Television         25188.60
Camera             24985.70
Air Conditioner    24812.05
Name: PurchasePrice, dtype: float64

#### *Analyze customer sentiments related to different products*

In [169]:
merge4 = pd.merge(reviews, products, on = "ProductID")
df = pd.DataFrame(merge4)

In [170]:
# Creating Product Reviews Pivot
Product_Reviews = df.pivot_table(index=['ProductName'], columns='Sentiment', values='ReviewID', aggfunc='count', sort='desc')
# Calculating Sentiment Score
Sentiment_Score = (Product_Reviews['Positive']*1)+(Product_Reviews['Neutral']*0)+(Product_Reviews['Negative']*-1)

In [171]:
# Creating Products Sentiment Score table
Product_Reviews_Score = Product_Reviews.assign(Sentiment_Score=Sentiment_Score)
Product_Reviews_Score
# Sorting the table by Sentiment Score 
Sorted_Product_Reviews = Product_Reviews_Score.sort_values(by='Sentiment_Score', ascending=False)
Sorted_Product_Reviews

Sentiment        Negative  Neutral  Positive  Sentiment_Score
ProductName                                                  
Smartwatch             10       19        27               17
Television             16       13        23                7
Heater                 13       25        19                6
Washing Machine        15       12        20                5
Printer                15       22        18                3
Toaster                17       23        20                3
Microwave              16       25        19                3
Smart Bulb             16       14        17                1
Blender                13        9        14                1
Air Conditioner        18       18        18                0
Laptop                 18       14        17               -1
Headphones             17       16        16               -1
Coffee Maker           13       22        12               -1
Camera                 17       15        16               -1
Smartphone             17       13        15               -2
Tablet                 20        9        18               -2
Electric Kettle        18       17        13               -5
Refrigerator           22       15        17               -5
Router                 21       11        14               -7
Vacuum Cleaner         19       15         9              -10

#### *Analyze customer sentiments related to different categories*

In [172]:
# Creating Category Reviews Pivot
Category_Reviews = df.pivot_table(index='ProductCategory', columns='Sentiment', values='ReviewID', aggfunc='count')
# Calculating Sentiment Score
Sentiment_Score = (Category_Reviews['Positive']*1)+(Category_Reviews['Neutral']*0)+(Category_Reviews['Negative']*-1)

In [173]:
# Creating Category Sentiment Score table
Category_Reviews_Score = Category_Reviews.assign(Sentiment_Score=Sentiment_Score)
Category_Reviews_Score
# Sorting the table by Sentiment Score 
Sorted_Category_Reviews = Category_Reviews_Score.sort_values(by='Sentiment_Score', ascending=False)
Sorted_Category_Reviews

Sentiment        Negative  Neutral  Positive  Sentiment_Score
ProductCategory                                              
Electronics           167      146       181               14
Home Appliances       164      181       161               -3

***
## **Summary Report**
***

### Assumptions ###
1. While analysing the data I have assumed that all the customers who made purchases are unique and have made purchase only once. So, accordingly I have
given unique IDs to all the customers in the dataset.

 *Reason: 
According to the data provided, even customers with the same name are from different countries. Because of that, I have assumed that they are from different customers with same name.*

2. While analysing the data I have assumed that the customers have given mixed reviews for all the different products, i.e. Positive, Neutral and Negative. And I have analyzed the customer sentiments based on those categories.

### Statistics

- ##### Total Purchased Quantity:
  3053

- ##### Total Revenue:
  489274.72

- ##### Average Purchase Value:
  489.27


### Top customers and their Purchase Behavior ###

- ##### Top Customers:
   Jillian Baker, Patricia Taylor, Mikayla Rios, Matthew Espinoza, Jamie Montoya, Charles Clark, Antonio Hughes, Carlos Smith,
                  Steven Rivera, Casey Johnson

- ##### Purchase Behavior:
  -  Electronics Category Dominance:  Most top customers purchased Electronics, indicating a preference for high-value and tech-related products
                                    like Smartphones, Cameras, and Headphones.
                                    
  -  High Purchase Frequency:  Customers like Patricia Taylor, Mikayla Rios, and Carlos Smith make multiple purchases over time, 
                                    suggesting they are regular buyers.
                                           
  -  Diverse Consumer Behavior:  They vary in terms of purchase frequency and product preferences, reflecting diverse consumer behaviors 
                                    within the Electronics and Home Appliances categories.

    
### Purchase Trends over Time ###

- ##### Monthly:
  The analysis suggests that consumer spending does not have much variations throughout the year, with a mild increase in March, April and Decemeber.

- ##### Quarterly:
  The consumer spending has been constant throughout the year, with a little increase in the second quarter. 
  Last quarter includes a slight dip  in consumer spending when compared to others.

- ##### Yearly:
  The yearly sales has a slight increase in the year 2024 based on the current given data.


### Top-Performing Product Categories ###

- ##### Vacuum Cleaner: 
  With a total reviews count of 2,48,194.56, the Electronics category emerges as the top-performing category based on sales. This category includes 
  products like Smartphones, Cameras, Televisions, and Headphones, which are typically higher in value and often have a higher profit margin.


### Top-Performing Products ###

  Consumers are spending more on luxury products like Microwave, Smartwatch and Toaster than some other necessary products like Smartphones 
  and Washing Machine.


### Sentiments over Products ###

- ##### Positive: 
  Smartwatch has the highest sentiment score with a count of 27 positive reviews, 19 neutral reviews and 10 negative reviews. This shows that Smartwatch
  is most liked by the customers.
  Some other products with good sentiment scores are Television, Heater and Washing Machine.

- ##### Negative: 
  Vaccum Cleaner has the lowest sentiment score with a count of 19 negative reviews, 15 neutral and 9 positive reviews, which indicates
  that, Vaccum Cleaner is the least liked product.
  Some other products with low sentiment scores are Router, Refrigerator and Electric Kettle.


### Sentiments over Categories ###

  Electronics has higher sentiment score than Home Appliances. The sentiments of customers for Home Appliances is mostly negative.
  
### **Conclusion and Key Insights** ###

- Electronics Category Dominance
- Diverse Consumer Behavior
- Luxury Products
- Smartwatch Preferability
- Vaccum Cleaner and Router Dislikement

Understanding these patterns can help in tailoring marketing strategies and product offerings to better serve the customer segments.
Businesses can use these insights to focus on optimizing their offerings, marketing strategies, and inventory management within the Electronics 
category and luxury products to capitalize on consumer preferences and maximize profitability throughout the year.

In [177]:
customers.to_excel('customers.xlsx',index=False)
products.to_excel('products.xlsx',index=False)
purchase.to_excel('purchase.xlsx',index=False)
reviews.to_excel('reviews.xlsx',index=False)
Product_Reviews.to_excel('Product_Reviews.xlsx')

In [204]:
Product_Reviews_Score.to_excel('Product_Reviews_Score.xlsx')
Category_Reviews_Score.to_excel('Category_Reviews_Score.xlsx')